syntax = "proto3"

package com.tbsen.proto;

// 공통 자료형
enum AgentState {
    STATE_UNKNOWN = 0;
    STATE_RUNNING = 1;
    STATE_WARNING = 2;
    STATE_ERROR = 3;
    STATE_STOPPED = 4;
}

// XDP 정의
enum XdpMode {
    MODE_UNKNOWN = 0;
    MODE_NATIVE = 1;
    MODE_SKB = 2;
    MODE_OFFLOAD = 3;
}

// 필터 백엔드 정의
enum FilterBackend {
  BACKEND_UNKNOWN = 0;
  BACKEND_XDP = 1;
  BACKEND_NFTABLES = 2;
}

enum ActionType {
    ACTION_UNKNOWN = 0;
    ACTION_ADD = 1;
    ACTION_DELETE = 2;
}

// ----- 메시지 정의 -----

// Agent 정보 등록
message AgentIdentity {
    string uuid = 1;
    string hostname =2;
    string host_ip = 3;
    string kernel_version = 4;
    string os_version = 5;
    string agent_version = 6;
    repeated string interfaces = 7; // 인터페이스 배열(리스트) 할당
}

message InterfaceMetrics {
    XdpMode current_mode = 1;
    uint64 pass_count = 2;
    uint64 drop_count = 3;
    double pps = 4;
}

// Agent Heartbeat
message AgentStatus {
    string uuid = 1;
    int64 timestamp = 2;
    AgentState global_state = 3;
    
    // 각 인터페이스별 지표
    map<string, InterfaceMetrics> interface_metrics = 4;

    // 리소스 상태
    double cpu_usage_percent = 5;
    double memory_usage_percent = 6;
}

// 룰 커맨드
message RuleCommand {
  string target_ip = 1;
  ActionType type = 2;
  string interface = 3; 
  FilterBackend backend_type = 4; // XDP / NFTABLES 
}

// 커맨드 응답
message CommandResponse {
    bool success = 1;
    string message = 2;
}

// RPC
service FilterAgent {

    rpc RegisterAgent (AgentIdentity) returns (CommandResponse);

    rpc ReportStatus (stream AgentStatus) returns (CommandResponse);

    rpc ManageRule (RuleCommand) returns (CommandResponse);
}